File: skills/building-calctree-calculations/examples/smoke-two-page.ts
/**
 * Two-page smoke test against a SANDBOX workspace. Run this first to confirm your
 * key and endpoint work end to end:
 *   1. create page A (+tree), give it inputs + a derived result
 *   2. read A back (getPageContext) and confirm it evaluated server-side
 *   3. create page B (+tree)
 *   4. reference A into B (a snapshot of A's values)
 *   5. read B back, print everything + the /edit URLs
 *
 * It creates two real pages, so point it at a workspace you do not mind writing to.
 *
 * Usage: CALCTREE_API_KEY=... npx tsx examples/smoke-two-page.ts <workspaceId>
 */
import {
  createPageInTree, insertMDXContent, getPageContext, referencePageViaApi,
} from '../scripts/calctree-api.ts'

const ws = process.argv[2]
if (!ws) { console.error('usage: npx tsx examples/smoke-two-page.ts <workspaceId>'); process.exit(1) }
if (!process.env.CALCTREE_API_KEY) { console.error('set CALCTREE_API_KEY'); process.exit(1) }
const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

console.log('using CALCTREE_API_KEY\n')
const url = (id: string) => `https://app.calctree.com/edit/${ws}/${id}`
const fmtVal = (raw: unknown): string => {
  let v: unknown = raw
  if (typeof v === 'string') { try { v = JSON.parse(v) } catch { return String(raw) } }
  if (v && typeof v === 'object' && (v as { mathjs?: string }).mathjs === 'Unit') {
    const u = v as { value: number; unit: string }; return `${u.value} ${u.unit}`
  }
  return JSON.stringify(v)
}
const show = (label: string, ctx: Awaited<ReturnType<typeof getPageContext>>) => {
  console.log(`  ${label}: ${ctx.statements.length} statements`)
  for (const s of ctx.statements) {
    const vals = s.namedValues.filter((v) => v.name).map((v) => `${v.name}=${fmtVal(v.value)}`).join(', ')
    const errs = s.errors?.length ? `  ERRORS: ${JSON.stringify(s.errors)}` : ''
    console.log(`    [${s.engine}] ${s.formula.replace(/\n/g, ' ')}${vals ? `  => ${vals}` : '  (no value)'}${errs}`)
  }
}

// 1. Page A with inputs + derived result
const A = await createPageInTree(ws, 'Smoke A — inputs')
console.log(`created A: ${url(A.id)}`)
// Prose AND formula both go through insertMDXContent. The EquationBlock carries the
// formula, so the body node and the persisted statement are created together.
// No leading '#' — the page title already renders as the H1.
const mdxA = [
  'Simply-supported beam: UDL over a clear span, mid-span moment.', '',
  '<EquationBlock name="Beam moment">', '```', 'span = 8 m', 'load = 45 kN / m', 'M_max = load * span^2 / 8', '```', '</EquationBlock>', '',
].join('\n')
const insA = await insertMDXContent(ws, A.id, mdxA)
console.log(`  A via insertMDXContent: ${insA.insertedCount} nodes, ${insA.statementsCreated} statements`)

// 2. read A back (settle first — statements evaluate async)
await sleep(2000)
const a = await getPageContext(ws, A.id)
show('page A', a)

// 3. Page B
const B = await createPageInTree(ws, 'Smoke B — references A')
console.log(`\ncreated B: ${url(B.id)}`)
await insertMDXContent(ws, B.id, 'Pulls a point-in-time snapshot of page A’s values (a Nodes-panel linked-page card, by design).')

// 4. reference A into B (stays a calc-graph "Page:" card — correct, not a body block)
const ref = await referencePageViaApi(ws, B.id, A.id)
console.log(`  referencePageViaApi -> alias="${ref.alias}", ${ref.count} values snapshotted`)

// 5. read B back (settle first)
await sleep(2000)
const b = await getPageContext(ws, B.id)
show('page B', b)

console.log(`\nDONE.\n  A: ${url(A.id)}\n  B: ${url(B.id)}`)

File: skills/calctree/examples/smoke_two_page.py
#!/usr/bin/env python3
"""
Two-page smoke test. Run this first to confirm your key and endpoint work end to end:

  1. create page A (registered in the tree) with inputs and a derived result
  2. read A back and confirm it evaluated server-side
  3. create page B
  4. reference A into B — a snapshot of A's computed values
  5. read B back and print both /edit URLs

It creates two real pages, so point it at a workspace you do not mind writing to.

Usage: CALCTREE_API_KEY=... python3 examples/smoke_two_page.py <workspaceId>
"""
import json
import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "scripts"))

from calctree_api import (  # noqa: E402
    CalcTreeError, create_page_in_tree, get_page_context, insert_mdx_content,
    apply_mdx_statement_titles, page_url, reference_page_via_api,
)

MDX_A = """Simply-supported beam: a UDL over a clear span, mid-span moment.

<EquationBlock name="Beam moment">
```
span = 8 m
load = 45 kN / m
M_max = load * span^2 / 8
```
</EquationBlock>
"""


def fmt(raw) -> str:
    v = raw
    if isinstance(v, str):
        try:
            v = json.loads(v)
        except (ValueError, TypeError):
            return str(raw)
    if isinstance(v, dict) and v.get("mathjs") == "Unit":
        val = v.get("value")
        if isinstance(val, float) and val.is_integer():
            val = int(val)
        return f"{val} {v.get('unit')}"
    return json.dumps(v, separators=(",", ":"))


def show(label: str, ctx: dict) -> None:
    print(f"  {label}: {len(ctx['statements'])} statement(s)")
    for s in ctx["statements"]:
        vals = ", ".join(f"{nv['name']}={fmt(nv.get('value'))}"
                         for nv in (s.get("namedValues") or []) if nv.get("name"))
        errs = f"  ERRORS: {s['errors']}" if s.get("errors") else ""
        formula = (s.get("formula") or "").replace("\n", " ")
        print(f"    [{s['engine']}] {s.get('title')!r} {formula}")
        print(f"      => {vals or '(no value yet)'}{errs}")


def main() -> int:
    if len(sys.argv) < 2:
        print("usage: python3 examples/smoke_two_page.py <workspaceId>", file=sys.stderr)
        return 2
    ws = sys.argv[1]
    if not os.environ.get("CALCTREE_API_KEY"):
        print("set CALCTREE_API_KEY", file=sys.stderr)
        return 2
    print("using CALCTREE_API_KEY\n")

    # 1. page A, with a formula that has to evaluate server-side
    a = create_page_in_tree(ws, "Smoke A - inputs")
    print(f"created A: {page_url(ws, a['id'])}")
    ins = insert_mdx_content(ws, a["id"], MDX_A)
    print(f"  insertMDXContent: {ins['insertedCount']} nodes, "
          f"{ins['statementsCreated']} statement(s)")
    if ins["statementsCreated"] == 0 and ins["insertedCount"] > 0:
        print("  FAILED: nodes inserted but no statements created — the page will not "
              "compute. Check CALCTREE_API_KEY.", file=sys.stderr)
        return 1

    titles = apply_mdx_statement_titles(ws, a["id"], MDX_A)
    print(f"  titles: {titles['titled']} set, verified={titles['verified']}, "
          f"attempts={titles['attempts']}")
    if not titles["verified"]:
        print("  WARNING: statement titles were not confirmed on the page", file=sys.stderr)

    # 2. read A back — evaluation is async, so settle first
    time.sleep(2)
    show("page A", get_page_context(ws, a["id"]))
    print()

    # 3 + 4. page B, referencing A rather than recomputing it
    b = create_page_in_tree(ws, "Smoke B - references A")
    print(f"created B: {page_url(ws, b['id'])}")
    ref = reference_page_via_api(ws, b["id"], a["id"])
    print(f"  reference: alias={ref['alias']!r}, {ref['count']} value(s) snapshotted")

    # 5. read B back
    time.sleep(2)
    show("page B", get_page_context(ws, b["id"]))

    print("\nDONE.")
    print(f"  A: {page_url(ws, a['id'])}")
    print(f"  B: {page_url(ws, b['id'])}")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except CalcTreeError as e:
        print(f"error: {e}", file=sys.stderr)
        sys.exit(1)
